### PR TITLE
activate sourcecode formatting

### PR DESCRIPTION
**Pull request by mwiede**
This will use formatting rules of `eclipse-java-google-style.xml` to format the source code in maven `process-sources` phase.
There should be only whitespace changes in java files.

Checkstyle Plugin is added, but not used by default, since current code would have 2640 violations against `Google-Style`.
```
$ mvn validate -Pcheckstyle
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.848 s
[INFO] Finished at: 2022-11-29T10:34:30Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.2.0:check (validate) on project jsch: You have 2640 Checkstyle violations. -> [Help 1]
```
Cleaning these up is an additional effort.

### CODE REVIEW

**Comment by norrisjeremy** [comment]
Have you considered using [fmt-maven-plugin](https://github.com/spotify/fmt-maven-plugin)?
It directly utilizes Google [google-jave-format](https://github.com/google/google-java-format) library to process sources.
We typically add it's check goal like this to our projects so that non-conforming code causes a build failure during the `verify` phase:
```
      <plugin>
        <groupId>com.spotify.fmt</groupId>
        <artifactId>fmt-maven-plugin</artifactId>
        <version>2.19</version>
        <executions>
          <execution>
            <goals>
              <goal>check</goal>
            </goals>
            <phase>verify</phase>
          </execution>
        </executions>
      </plugin>
```

And to directly use it's format goal to reformat sources, you can manually execute `mvn fmt:format`.
It may require some additional tweaks to the plugin configuration to pick-up all the non-standard source directories where we put the > Java 8 source code.

**Comment by mwiede** [comment]
i have not compared fmt-maven-plugin, but as far as i can see formatter-maven-plugin offers the same. you can both format source code as well as verify it.

**Comment by norrisjeremy** [comment]
It appears `formatter-maven-plugin` also has a `validate` goal: is there a reason to to use `maven-checkstyle-plugin` instead of that?

Also, do we need to configure these plugins to also handle the > Java 8 code directories (`src/main/java9`, `src/main/java11`, etc.)?

File: src/main/java/com/jcraft/jsch/AgentConnector.java [hunk 1]
```diff
@@ -1,36 +1,36 @@
 /* -*-mode:java; c-basic-offset:2; indent-tabs-mode:nil -*- */
```
Comment (norrisjeremy) [inline src/main/java/com/jcraft/jsch/AgentConnector.java:1]: I think we may want to also remove all these editor hints at the top of all the files too?

File: pom.xml [hunk 2]
```diff
@@ -618,5 +618,60 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>format</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
```
Comment (norrisjeremy) [inline pom.xml:624]: Are we sure we want reformatting to be `activeByDefault`?
Or should it be that `maven-checkstyle-plugin` is `activeByDefault`?

Comment (mwiede) [reply pom.xml:624]: if we run formatting by default, there is not need to run checkstyle. It's just optional to make it available, but at the moment, the result of reformatting does not remove any of the checkstyle violations.

File: src/main/java/com/jcraft/jsch/AgentConnector.java (hunk 1 above)
Comment (mwiede) [reply src/main/java/com/jcraft/jsch/AgentConnector.java:1]: yes makes sense

File: pom.xml [hunk 3]
```diff
@@ -618,5 +618,71 @@
...
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-checkstyle-plugin</artifactId>
+                        <version>3.2.0</version>
+                        <configuration>
+                            <encoding>UTF-8</encoding>
+                            <failOnViolation>true</failOnViolation>
+                            <logViolationsToConsole>true</logViolationsToConsole>
+                            <violationSeverity>warning</violationSeverity>
+                            <configLocation>google_checks.xml</configLocation>
+                        </configuration>
```
Comment (norrisjeremy) [inline pom.xml:673]: Does `maven-checkstyle-plugin` need configuration for the non-standard source directories as well?

File: pom.xml (hunk 2 above)
Comment (norrisjeremy) [reply pom.xml:624]: Does checkstyle report violations even after formatter runs and adjusts everything though?

Comment (norrisjeremy) [reply pom.xml:624]: Ok, I see now that checkstyle logs a huge number of warnings and the build fails when it is activated, so I see now why you didn't activate it by default.

File: pom.xml [hunk 4]
```diff
@@ -618,5 +646,64 @@
...
             </build>
         </profile>
+        <profile>
+            <id>format</id>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>net.revelc.code.formatter</groupId>
+                        <artifactId>formatter-maven-plugin</artifactId>
+                        <version>2.21.0</version>
```
Comment (norrisjeremy) [inline pom.xml:656]: I suspect the `<version>` doesn't need to be specified here, as it should pick it up from earlier?

**Review by norrisjeremy — COMMENTED** [review]
Do we know what's causing the test failures during CI builds?

**Comment by mwiede** [comment]
> Do we know what's causing the test failures during CI builds?

I looked at it. The problem seems to be the docker bind mount of the temporary directory. We need to move the temp directory to the current workspace.

**Comment by mwiede** [comment]
> > Do we know what's causing the test failures during CI builds?
> 
> I looked at it. The problem seems to be the docker bind mount of the temporary directory. We need to move the temp directory to the current workspace.

that was a problem I had within Github Codespaces environment. The other problem was not resolved by downgrading the runner image to ubuntu 20.04

File: pom.xml (hunk 4 above)
Comment (mwiede) [reply pom.xml]: removed it

File: src/main/java/com/jcraft/jsch/AgentConnector.java (hunk 1 above)
Comment (norrisjeremy) [reply src/main/java/com/jcraft/jsch/AgentConnector.java:1]: Did we want to handle removal of these editor hints in this PR or do that separately

Comment (mwiede) [reply src/main/java/com/jcraft/jsch/AgentConnector.java:1]: I will add it to this one

File: pom.xml [hunk 5]
```diff
@@ -509,6 +509,26 @@
                 <artifactId>maven-dependency-plugin</artifactId>
                 <version>3.5.0</version>
             </plugin>
+            <plugin>
+                <groupId>net.revelc.code.formatter</groupId>
+                <artifactId>formatter-maven-plugin</artifactId>
+                <version>2.21.0</version>
```
Comment (norrisjeremy) [inline pom.xml:515]: 2.22.0 is the newest version.

File: pom.xml [hunk 6]
```diff
@@ -612,5 +632,84 @@
...
+            </build>
+        </profile>
+        <profile>
+            <id>checkstyle</id>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-checkstyle-plugin</artifactId>
+                        <version>3.2.0</version>
```
Comment (norrisjeremy) [inline pom.xml:682]: 3.2.1 is the newest version.

File: pom.xml [hunk 7]
```diff
@@ -612,5 +632,84 @@
...
+                                    <goal>format</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+            <activation>
```
Comment (norrisjeremy) [inline pom.xml:655]: We should add an explicit `<id>` for this profile.

**Comment by norrisjeremy** [comment]
I think we may want to add `-Dformatter.skip=true` in `.github/actions/maven.yml` for the `Test with Maven` step.
I believe that will prevent Github CI from executing the formatter validation check when it executes the unit & integration tests (since it already would have done this during the prior `Build with Maven` step).

**Comment by norrisjeremy** [comment]
> I think we may want to add `-Dformatter.skip=true` in `.github/workflows/maven.yml` for the `Test with Maven` step.
> I believe that will prevent Github CI from executing the formatter validation check when it executes the unit & integration tests (since it already would have done this during the prior `Build with Maven` step).

Disregard, I missed that formatter:validate is tied to verify phase, so it has to happen during the `Test with Maven` step.

File: pom.xml (hunk 5 above)
Comment (mwiede) [reply pom.xml]: updated

File: pom.xml (hunk 6 above)
Comment (mwiede) [reply pom.xml]: updated

File: pom.xml (hunk 7 above)
Comment (mwiede) [reply pom.xml:656]: 👍

File: pom.xml [hunk 8]
```diff
@@ -612,5 +632,85 @@
...
+                                    <goal>format</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+			<id>formatvalidation</id>
```
Comment (norrisjeremy) [inline pom.xml:655]: I think the indentation is off here?

**Review by norrisjeremy — APPROVED** [review]
I think we should do a squash merge of this so it's a single commit.

**Pull request merged**